src/interfaces/process.jl

```diff
@@ -191,15 +191,48 @@ Convenience function dispatching to [`incoming_particles`](@ref) or [`outgoing_p
     outgoing_particles(proc_def)
 
 """
-    number_particles(proc_def::AbstractProcessDefinition, ::ParticleDirection)
+    number_particles(proc_def::AbstractProcessDefinition, dir::ParticleDirection)
 
-Convenience function dispatching to [`number_incoming_particles`](@ref) or [`number_outgoing_particles`](@ref) depending on the given direction.
+Convenience function dispatching to [`number_incoming_particles`](@ref) or [`number_outgoing_particles`](@ref) depending on the given direction, returning the number of incoming or outgoing particles, respectively.
 """
 @inline number_particles(proc_def::AbstractProcessDefinition, ::Incoming) =
     number_incoming_particles(proc_def)
 @inline number_particles(proc_def::AbstractProcessDefinition, ::Outgoing) =
     number_outgoing_particles(proc_def)
 
+"""
+    number_particles(proc_def::AbstractProcessDefinition, dir::ParticleDirection, species::AbstractParticleType)
+
+Return the number of particles of the given direction and species in the given process definition.
+"""
+@inline function number_particles(
+    proc_def::AbstractProcessDefinition, dir::DIR, ::PT
+) where {DIR<:ParticleDirection,PT<:AbstractParticleType}
+    return count(x -> x isa PT, particles(proc_def, dir))
+end
+
+"""
+    number_particles(proc_def::AbstractProcessDefinition, particle::AbstractParticleStateful)
+    number_particles(proc_def::AbstractProcessDefinition, particle_type::Type{AbstractParticleStateful})
+
+Return the number of particles of the given particle's direction and species in the given process definition.
+"""
+@inline function number_particles(
+    proc_def::AbstractProcessDefinition, ::AbstractParticleStateful{DIR,PT,EL}
+) where {DIR<:ParticleDirection,PT<:AbstractParticleType,EL<:AbstractFourMomentum}
+    return number_particles(proc_def, DIR(), PT())
+end
+@inline function number_particles(
```

Comment:
Is this really needed? Having functions that operate on objects and types at the same time is somewhat dangerous. 
Again, the call of the type parameter for dispatch smells a bit. However, I think it is the only option here unless we implement `number_particle` on the types of the species and direction as well. At the end of the day, this would not be a good practice as you can find here: https://docs.julialang.org/en/v1/manual/style-guide/#Avoid-confusion-about-whether-something-is-an-instance-or-a-type

Reply:
The problem I was trying to solve here was that sometimes I wanted to know how many particles exist  of some direction and species, but I don't have an initialized `ParticleStateful` at hand. Since it has state (i.e. a momentum), I wouldn't want to initialize it first, just to use it in this function call where the momentum is irrelevant anyways.

Thinking about it now though the better option would be to simply provide more basic overloads of `number_particles` that just take the direction and species instead of a full particle stateful. The current implementation on a `ParticleStateful` can then just become a dispatch on that. That seems much cleaner.

Reply:
I further realize now that that's actually what I already did :sweat_smile: 
I'll just delete the function.